aten/src/ATen/core/ivalue.cpp

```diff
@@ -13,8 +13,63 @@ CAFFE2_API c10::intrusive_ptr<ConstantString> ConstantString::create(
   return c10::make_intrusive<ConstantString>(std::move(str_));
 }
 
+TupleTypePtr Tuple::type() const {
+  if (!type_) {
+    type_ = TupleType::create(
+        fmap(elements_, [&](const IValue& v) { return v.type(); }));
+  }
+  return type_;
+}
+
 } // namespace ivalue
 
+
+TypePtr IValue::type() const {
+  switch(tag) {
+    case Tag::None:
+      return NoneType::get();
+    case Tag::Tensor:
+      return TensorType::create(toTensor());
+    case Tag::Double:
+      return FloatType::get();
+    case Tag::Int:
+      return IntType::get();
+    case Tag::Bool:
+      return BoolType::get();
+    case Tag::IntList:
+      return ListType::ofInts();
+    case Tag::DoubleList:
+      return ListType::ofFloats();
+    case Tag::BoolList:
+      return ListType::ofBools();
+    case Tag::TensorList:
+      return ListType::ofTensors();
+    case Tag::String:
+      return StringType::get();
+    case Tag::Blob:
+      return AnyType::get();
+    case Tag::GenericDict: {
+      auto d = toGenericDict();
+      return DictType::create(d.keyType(), d.valueType());
+    }
+    case Tag::GenericList:
+      return ListType::create(toGenericList().elementType());
+    case Tag::Future:
+      return toFuture()->type();
+    case Tag::Device:
+      return DeviceObjType::get();
+    case Tag::Object:
+      return toObjectRef().type();
+    case Tag::Uninitialized:
+      return AnyType::get();
```

Comment:
Aside: what is `Uninitialized` anyway? I'm asking because IValues that are `Any` type should be scrutinized

Reply:
It is introduce when we have early returns, break, and continue. It represents an uninitialized IValue that we know because of invariants introduced in the pass cannot be used. This fact isn't representable in the graph directly so this placeholder is needed. The alternative, representing them with optional values, requires a lot of unwrapping logic around use sites.